### PR TITLE
fix(resolve): only disambiguate binding key during define

### LINKED ISSUE #110164: ICE: stack overflow in resolver
### Code

```Rust
use crate::*;

fn main() {
    use _::*;
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.70.0-nightly (696aaad58 2023-04-09)
binary: rustc
commit-hash: 696aaad58c57a589f6fb2ecff5bae2eec581cb71
commit-date: 2023-04-09
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 16.0.2
```

### Error output

```
<output>
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: expected identifier, found reserved identifier `_`
 --> /tmp/treereduce.out:4:9
  |
4 |     use _::*;
  |         ^ expected identifier, found reserved identifier


Thread 2 "rustc" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe9dff6c0 (LWP 3656741)]
0x00007ffff534e571 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
(gdb) bt
#0  0x00007ffff534e571 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#1  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#2  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#3  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#4  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#5  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#6  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#7  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#8  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#9  0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#10 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#11 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#12 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#13 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#14 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#15 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#16 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#17 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#18 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#19 0x00007ffff534fcb0 in <rustc_resolve[a2234260a32b12fb]::Resolver>::resolve_ident_in_module_unadjusted_ext () from /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-e24c3287901edaff.so
#20 0


.....
```

</p>
</details>

### PR DESCRIPTION
**Pull request by bvanjoi**
- close https://github.com/rust-lang/rust/issues/110164
- discussion: https://github.com/rust-lang/rust/pull/110264#discussion_r1168895468

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_resolve/src/build_reduced_graph.rs [hunk 1]
```diff
@@ -9,7 +9,9 @@ use crate::def_collector::collect_definitions;
 use crate::imports::{Import, ImportKind};
 use crate::macros::{MacroRulesBinding, MacroRulesScope, MacroRulesScopeRef};
 use crate::Namespace::{self, MacroNS, TypeNS, ValueNS};
-use crate::{Determinacy, ExternPreludeEntry, Finalize, Module, ModuleKind, ModuleOrUniformRoot};
+use crate::{
+    BindingKey, Determinacy, ExternPreludeEntry, Finalize, Module, ModuleKind, ModuleOrUniformRoot,
+};
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/build_reduced_graph.rs:14]: Nit: cold you keep the single line import style everywhere?

**Comment by bvanjoi** [comment]
It may be necessary to approve again, probably because the subsequent `push --force` has left the queue in an unapproved state @petrochenkov

**Comment by bvanjoi** [comment]
Clearing git conflicts completed, @rustbot ready

**Comment by wesleywiser** [comment]
Manually reviewed the diffs and the code changes are the same as when @petrochenkov originally reviewed modulo fixes to use statements during rebase. 

@bors r=petrochenkov

**Pull request merged**